### PR TITLE
Fix allocation of 0-length array

### PR DESCRIPTION
**Pull request by thebrid**
## Description
Using Valgrind to troubleshoot a crash, came across this. Not the crash I was looking for but interesting nonetheless. If we're called with argc=0, we allocate a 0-length array, pass it to Python then we get a memory error as Python tries to access member 0 of the 0-length array.

## Motivation and Context
Fixes an invalid memory read.

## How Has This Been Tested?
I clicked around a bunch and didn't manage to get it to crash. Re-ran in valgrind and the invalid memory read is no longer there. Ran all existing tests.

## Types of change
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [X] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [X] All new and existing tests passed

### CODE REVIEW

**Review by Paxxi — APPROVED** [review]
Looks good to me

File: xbmc/interfaces/python/PythonInvoker.cpp [hunk 1]
```diff
@@ -147,7 +147,12 @@ bool CPythonInvoker::execute(const std::string &script, const std::vector<std::s
 
   // copy the arguments into a local buffer
   m_argc = arguments.size();
-  m_argv = new char*[m_argc];
+
+  if(m_argc > 0)
```
Comment (garbear) [inline xbmc/interfaces/python/PythonInvoker.cpp]: missing whitespace

File: xbmc/interfaces/python/PythonInvoker.cpp [hunk 2]
```diff
@@ -147,7 +147,12 @@ bool CPythonInvoker::execute(const std::string &script, const std::vector<std::s
...
+  {
+    m_argv = new char*[m_argc];
+  }
```
Comment (garbear) [inline xbmc/interfaces/python/PythonInvoker.cpp]: also the braces aren't required for only 1 line

**Comment by garbear** [comment]
other than the code style comments it looks good

File: xbmc/interfaces/python/PythonInvoker.cpp [hunk 3]
```diff
@@ -147,7 +147,12 @@ bool CPythonInvoker::execute(const std::string &script, const std::vector<std::s
 
   // copy the arguments into a local buffer
   m_argc = arguments.size();
-  m_argv = new char*[m_argc];
+
+  if(m_argc > 0)
+  {
+    m_argv = new char*[m_argc];
```
Comment (garbear) [inline xbmc/interfaces/python/PythonInvoker.cpp]: this will leak if CPythonInvoker::execute() is called twice

File: xbmc/interfaces/python/PythonInvoker.cpp (hunk 2 above)
Comment (thebrid) [inline xbmc/interfaces/python/PythonInvoker.cpp]: Is it OK if I keep them? I've seen people in the past attempt to add new lines to conditional blocks without braces so I try to add braces as a rule.

Comment (Paxxi) [inline xbmc/interfaces/python/PythonInvoker.cpp]: Extra braces are fine unless they severely reduce readability so keep them

File: xbmc/interfaces/python/PythonInvoker.cpp (hunk 3 above)
Comment (thebrid) [inline xbmc/interfaces/python/PythonInvoker.cpp]: I figured this wasn't a regression so I didn't change it. I don't mind making the change though. Will send another commit in a little while.

File: xbmc/interfaces/python/PythonInvoker.cpp [hunk 4]
```diff
@@ -95,9 +97,27 @@ static const std::string getListOfAddonClassesAsString(XBMCAddon::AddonClass::Re
...
+	return output;
+}
+
+static std::vector<char *> convertVectorsToCharPtrs(std::vector<std::vector<char>> & input)
+{
+	std::vector<char *> output;
+	std::transform(input.begin(), input.end(), std::back_inserter(output), [](std::vector<char> & i) { return &i[0]; });
+	return output;
+}
+
```
Comment (wsnipex) [inline xbmc/interfaces/python/PythonInvoker.cpp]: indentation on the whole block is off. We use 2 spaces.

Comment (thebrid) [inline xbmc/interfaces/python/PythonInvoker.cpp]: :+1:

**Comment by thebrid** [comment]
@garbear feedback implemented
@wsnipex indentation should be fixed now

File: xbmc/interfaces/python/PythonInvoker.cpp [hunk 5]
```diff
@@ -147,7 +147,12 @@ bool CPythonInvoker::execute(const std::string &script, const std::vector<std::s
 
   // copy the arguments into a local buffer
   m_argc = arguments.size();
-  m_argv = new char*[m_argc];
+  
+  if(m_argc > 0)
+  {
+    m_argv = new char*[m_argc];
+  }
```
Comment (garbear) [inline xbmc/interfaces/python/PythonInvoker.cpp]: This block doesn't fix the cosmetic issues I mentioned earlier

File: xbmc/interfaces/python/PythonInvoker.cpp [hunk 6]
```diff
@@ -114,12 +134,6 @@ CPythonInvoker::~CPythonInvoker()
   Stop(true);
   pulseGlobalEvent();
 
-  if (m_argv != NULL)
-  {
-    for (unsigned int i = 0; i < m_argc; i++)
-      delete [] m_argv[i];
-    delete [] m_argv;
-  }
```
Comment (garbear) [inline xbmc/interfaces/python/PythonInvoker.cpp]: Replacing the member variables with local ones is the correct fix. However, you replace some overly-complex c++ code with some overly-complicated std container code. I recommend splitting this into two commits, one that moves this block  into CPythonInvoker::execute() and replaces the members with locals (keeping the c++ memory management), and then one that converts from `std::vector<std::string>` to `char**`.

**Comment by thebrid** [comment]
@garbear changed commits as you requested.

**Comment by Paxxi** [comment]
jenkins build this please

**Comment by wsnipex** [comment]
jenkins build this please

**Comment by DaveTBlake** [comment]
@thebrid check forum PM for using valgrind on another Kodi issue.

**Comment by Rechi** [comment]
If you refer to Travis CI, we don't care about it and it also fails a bit randomly. Only jenkins must be successful.

**Comment by wsnipex** [comment]
@garbear ping

**Review by garbear — APPROVED** [review]

**Comment by wsnipex** [comment]
thanks for the patience :)

**Comment by MilhouseVH** [comment]
Error in youtube plugin: http://sprunge.us/UZBD

Reverting this PR fixes the error.

**Comment by thebrid** [comment]
@MilhouseVH checking
